keras_cv/models/densenet.py

```diff
@@ -0,0 +1,296 @@
+    """A transition block.
+
+    Args:
+      x: input tensor.
+      reduction: float, compression rate at transition layers.
+      name: string, block label.
+
+    Returns:
+      output tensor for the block.
+    """
+    x = layers.BatchNormalization(axis=BN_AXIS, epsilon=1.001e-5, name=name + "_bn")(x)
+    x = layers.Activation("relu", name=name + "_relu")(x)
+    x = layers.Conv2D(
+        int(backend.int_shape(x)[BN_AXIS] * reduction),
+        1,
+        use_bias=False,
+        name=name + "_conv",
+    )(x)
+    x = layers.AveragePooling2D(2, strides=2, name=name + "_pool")(x)
+    return x
+
+
+def conv_block(x, growth_rate, name):
+    """A building block for a dense block.
+
+    Args:
+      x: input tensor.
+      growth_rate: float, growth rate at dense layers.
+      name: string, block label.
+
+    Returns:
+      Output tensor for the block.
+    """
+    x1 = layers.BatchNormalization(axis=BN_AXIS, epsilon=1.001e-5, name=name + "_0_bn")(
+        x
+    )
+    x1 = layers.Activation("relu", name=name + "_0_relu")(x1)
+    x1 = layers.Conv2D(4 * growth_rate, 1, use_bias=False, name=name + "_1_conv")(x1)
+    x1 = layers.BatchNormalization(axis=BN_AXIS, epsilon=1.001e-5, name=name + "_1_bn")(
+        x1
+    )
+    x1 = layers.Activation("relu", name=name + "_1_relu")(x1)
+    x1 = layers.Conv2D(
+        growth_rate, 3, padding="same", use_bias=False, name=name + "_2_conv"
+    )(x1)
+    x = layers.Concatenate(axis=BN_AXIS, name=name + "_concat")([x, x1])
+    return x
+
+
+def _get_name(blocks):
+    if blocks == [6, 12, 24, 16]:
+        return "densenet121"
+    if blocks == [6, 12, 32, 32]:
+        return "densenet169"
+    if blocks == [6, 12, 48, 32]:
+        return "densenet201"
+    return "densenet"
+
+
+def _apply_classifier(x, classes, classifier_activation):
+    x = layers.GlobalAveragePooling2D(name="avg_pool")(x)
+    x = layers.Dense(classes, activation=classifier_activation, name="predictions")(x)
+    return x
+
+
+def _apply_pooling_layer(x, pooling):
+    print(x.shape)
+    if pooling == "avg":
+        return layers.GlobalAveragePooling2D(name="avg_pool")(x)
+    if pooling == "max":
+        return layers.GlobalMaxPooling2D(name="max_pool")(x)
+    return x
+
+
+def DenseNet(
+    blocks,
+    include_rescaling,
+    include_top,
+    classes=None,
+    weights=None,
+    input_shape=(None, None, 3),
+    pooling=None,
+    classifier_activation="softmax",
+):
+    """Instantiates the DenseNet architecture.
+
+    Reference:
+    - [Densely Connected Convolutional Networks (CVPR 2017)](
+        https://arxiv.org/abs/1608.06993)
+
+    This function returns a Keras DenseNet model.
+
+    For transfer learning use cases, make sure to read the
+    [guide to transfer learning & fine-tuning](
+      https://keras.io/guides/transfer_learning/).
+
+    Args:
+      blocks: numbers of building blocks for the four dense layers.
+      include_rescaling: whether or not to Rescale the inputs.
```

Comment:
@LukeWood just an idea I got from our conversation offline. 

If we expose two additional _optional_ arguments called `mean` and `variance` we could just ignore whatever passed to `include_rescaling` and then promote the preprocessing to pass through a `Normalization(mean, variance)` layer. With sufficient documentation (like examples of what `mean` and `variance` should be etc.), it's possible to reduce the cognitive load of the user.

Reply:
Thanks for the suggestion, but let's not do that. We should universally standardize on a single preprocessing scheme and avoid customizability in this case. Standardization is what reduces cognitive load here.

As a user, if you want to do non-standard things, you can always use `False` then add the layers you want before the model.

Reply:
Makes sense.